internal/trace/listener.go

```diff
@@ -11,19 +11,144 @@ package trace
 import (
 	"context"
 	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/DataDog/datadog-lambda-go/internal/logger"
+	"github.com/aws/aws-lambda-go/lambdacontext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
 )
 
 type (
-	// Listener implements HandlerListener, injecting datadog tracing info into the context
-	Listener struct{}
+	// Listener creates a function execution span and injects it into the context
+	Listener struct {
+		DDTraceEnabled  bool
+		MergeXrayTraces bool
+	}
+
+	// Config gives options for how the Listener should work
+	Config struct {
+		DDTraceEnabled  bool
+		MergeXrayTraces bool
+	}
 )
 
-// HandlerStarted adds trace metadata to the context
+// The function execution span is the top-level span representing the current Lambda function execution
+var functionExecutionSpan ddtrace.Span
+
+// MakeListener initializes a new trace lambda Listener
+func MakeListener(config Config) Listener {
+
+	return Listener{
+		DDTraceEnabled:  config.DDTraceEnabled,
+		MergeXrayTraces: config.MergeXrayTraces,
+	}
+}
+
+// HandlerStarted creates the function execution span representing the Lambda function execution
+// and adds that span to the context so that the user can create child spans (if Datadog tracing is enabled)
 func (l *Listener) HandlerStarted(ctx context.Context, msg json.RawMessage) context.Context {
-	ctx, _ = ExtractTraceContext(ctx, msg)
+	ctx, _ = ContextWithTraceContext(ctx, msg)
+
+	if l.DDTraceEnabled {
+		tracer.Start(
+			tracer.WithService("aws.lambda"),
+			tracer.WithLambdaMode(true),
+			tracer.WithDebugMode(true),
+			tracer.WithGlobalTag("__dd.origin", "lambda"),
+		)
+
+		startFunctionExecutionSpan(
+			ctx,
+			l.MergeXrayTraces,
+		)
+
+		// Add the span to the context so the user can create child spans
+		ctx = tracer.ContextWithSpan(ctx, functionExecutionSpan)
+	}
+
 	return ctx
 }
 
-// HandlerFinished implemented as part of the HandlerListener interface
+// HandlerFinished finishes the function execution span (if it was started) and stops the tracer
 func (l *Listener) HandlerFinished(ctx context.Context) {
+	if functionExecutionSpan != nil {
+		functionExecutionSpan.Finish()
+	}
+
+	tracer.Stop()
```

Comment:
Does this need to be called every invocation?

Reply:
Per a conversation with @knusbaum, this is the preferred way to ensure that no state is maintained between invocations.

Although, maybe it is unnecessary? It does seem that the options we pass to the tracer when it starts are always the same (i.e. they don't depend on anything invocation-specific). Maybe it would be better to initialize the tracer outside of the listener and just leave it running?

Reply:
`tracer.Stop()` is currently the only way to force the tracer to flush any traces it's holding. If the execution shuts down after `HandlerFinished`, then we could drop traces. 

We could solve this in the future with a `Flush()` api call, or somesuch.

Reply:
I will add a comment documenting that